### PR TITLE
Issue 887/switch to std::atomic and std::mutex from deprecated tbb versions

### PR DESCRIPTION
**Pull request by BenFrantzDale**
Should I keep the commit that removes these `using` declarations,
```
using Mutex = std::mutex;	
using Lock = std::lock_guard<Mutex>;	
```
in favor of directly using the std types? Or would you like to keep these `using` declarations?

### CODE REVIEW

File: openvdb/openvdb/viewer/Viewer.cc [hunk 1]
```diff
@@ -328,7 +328,9 @@ ThreadManager::doView()
     // This function runs in its own thread.
     // The mClose and mRedisplay flags are set from the main thread.
     while (!mClose) {
-        if (mRedisplay.compare_and_swap(/*set to*/false, /*if*/true)) {
+        bool expected = true;
+	mRedisplay.compare_exchange_strong(expected, false);
+        if (expected) {
```
Comment (BenFrantzDale) [inline openvdb/openvdb/viewer/Viewer.cc:333]: This was the one tricky part, since the argument order and semantics are different. 

The tbb version is https://www.threadingbuildingblocks.org/docs/help/tbb_userguide/Atomic_Operations.html:
* `x.compare_and_swap(y,z)`: if `x` equals `z`, then do `x=y`. In either case, return old value of `x`.
The std version https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange does
* Atomically compares the object representation (until C++20)value representation (since C++20) of `*this` with that of `expected`, and if those are bitwise-equal, replaces the former with desired (performs read-modify-write operation). Otherwise, loads the actual value stored in `*this` into `expected` (performs load operation).
* Returns `true` if the underlying atomic value was successfully changed, `false` otherwise.

**Comment by Idclip** [comment]
This looks great @BenFrantzDale, thanks! Could you fix the dco signoff and the conflicts so the CI starts running?

**Comment by BenFrantzDale** [comment]
> This looks great @BenFrantzDale, thanks! Could you fix the dco signoff and the conflicts so the CI starts running?

Will do. I’ll also do tbb::mutex which is also going away.

The one other future-tbb-comparability issue I know of is `Queue.cpp` using `tbb::task`. I’m not sure how to address that one trivially.

**Comment by BenFrantzDale** [comment]
> > This looks great @BenFrantzDale, thanks! Could you fix the dco signoff and the conflicts so the CI starts running?
> 
> Will do. I’ll also do tbb::mutex which is also going away.
> 
> The one other future-tbb-comparability issue I know of is `Queue.cpp` using `tbb::task`. I’m not sure how to address that one trivially.

Along those lines, there are a lot of
```
using Mutex = tbb::mutex;
using Lock = Mutex::scoped_lock;
```
which I can change to 
```
using Mutex = std::mutex;
using Lock = std::lock_guard<Mutex>;
```
but since this is a move to the standard, is there any reason not to just drop those `using`s and use `std::mutex` and `std::lock_guard` in the code?

File: openvdb/openvdb/viewer/Viewer.cc [hunk 2]
```diff
@@ -328,7 +328,9 @@ ThreadManager::doView()
     // This function runs in its own thread.
     // The mClose and mRedisplay flags are set from the main thread.
     while (!mClose) {
-        if (mRedisplay.compare_and_swap(/*set to*/false, /*if*/true)) {
+        bool expected = true;
+        mRedisplay.compare_exchange_strong(expected, false);
+        if (expected) {
```
Comment (jmlait) [inline openvdb/openvdb/viewer/Viewer.cc:333]: Looks correct, but I 'm curious why not
if (mRedisplay.compare_exchange_strong(expected, false))
as it will return true on a successful store; which will correspond to the old behaviour?

**Review by jmlait — APPROVED** [review]

File: openvdb/openvdb/openvdb.cc [hunk 3]
```diff
@@ -130,14 +134,13 @@ void
...
     Lock lock(sInitMutex);
-
 #ifdef __ICC
 // Disable ICC "assignment to statically allocated variable" warning.
 // This assignment is mutex-protected and therefore thread-safe.
 __pragma(warning(disable:1711))
 #endif
 
-    sIsInitialized = false;
+    sIsInitialized.store(false, std::memory_order_seq_cst); // Do we need full memory order?
```
[the remark this answers is not in the document]
Comment (BenFrantzDale) [reply openvdb/openvdb/openvdb.cc:143]: OK, then I'll probably drop this comment. I agree, it's pretty sketchy.

File: openvdb/openvdb/openvdb.cc [hunk 4]
```diff
@@ -30,20 +33,21 @@ namespace openvdb {
...
 // Declare this at file scope to ensure thread-safe initialization.
 Mutex sInitMutex;
-bool sIsInitialized = false;
+std::atomic<bool> sIsInitialized{false};
 }
 
 void
 initialize()
 {
+    if (sIsInitialized.load(std::memory_order_acquire)) return;
```
[the remark this answers is not in the document]
Comment (BenFrantzDale) [reply openvdb/openvdb/openvdb.cc:48]: I think (?) we want at least acquire semantics, otherwise we could have an out-of-date read (right?) and decide that it's initialized when it's in fact no longer initialized? (Although tearing down seems sketchy to me.)

File: openvdb/openvdb/points/AttributeArray.cc [hunk 5]
```diff
@@ -67,7 +67,7 @@ AttributeArray::AttributeArray(const AttributeArray& rhs)
     : mIsUniform(rhs.mIsUniform)
     , mFlags(rhs.mFlags)
     , mUsePagedRead(rhs.mUsePagedRead)
-    , mOutOfCore(rhs.mOutOfCore)
+    , mOutOfCore(rhs.mOutOfCore.load())
```
[the remark this answers is not in the document]
Comment (BenFrantzDale) [reply openvdb/openvdb/points/AttributeArray.cc:70]: I think this line isn't a semantic change. I'm not sure what you are pointing out.

File: openvdb/openvdb/viewer/Viewer.cc (hunk 2 above)
Comment (BenFrantzDale) [reply openvdb/openvdb/viewer/Viewer.cc:333]: tbb says  "x.compare_and_swap(y,z) | if x equals z, then do x=y. In either case, return old value of x." https://www.threadingbuildingblocks.org/docs/help/tbb_userguide/Atomic_Operations.html
So in the old version...if `mRedisplay` ....
* was `false` it won't be swapped, so will return `false`.
* was `true` it will be swapped so it will return `true`.
New version:
If it...
* was `false` it won't be swapped (because `expected` is `true`) and so will return `false` and `expected` == `false.
* was `true` it will be swapped and so will return `true` and `expected == true`
So yes, you are right. I was pedantically substituting such that the `if` is checking against the old value of `mRedisplay` just like it always did. But yes, in this case `expected` is always equal to the result of `compare_exchange_strong`. I'd go either way here. Your version is more concise but is saying "Do a compare_exchange_strong and if that changed it...." versus my "Do a compare_exchange_strong and if `mRedisplay` was `true`...". I don't know which is clearer.

File: openvdb/openvdb/openvdb.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (BenFrantzDale) [reply openvdb/openvdb/openvdb.cc:48]: The only approach here that I'd feel confident in that would still permit `uninitialize` would be to have the shared state that goes with initialization be stored attached to a in a `static std::shared_ptr` and have initialization atomically get it, check it and return a `std::shared_ptr` to it, making it the caller's job to keep the `shared_ptr` alive.

As it is, I think the caller of `uninitialize` always has to worry about if anyone is still using things. If `uninitialize` weren't needed, we could just use `std::once_flag`.

[the remark this answers is not in the document]
Comment (jmlait) [reply openvdb/openvdb/openvdb.cc:48]: This is what I meant about uninitialize being sketchy :>

Acquire doesn't stop you from having an out-of-date read.  You can still have an out of date read as there is a window between when you test this and the double check lock.  So you can read this with acquire; properly validate that it is initialized;  have else-thread unintialize; and then return as if initialized.

As soon as you have uninitialize interleaving with initialize this is all quite broken, as there is no sense of a stack.  So we might as well make the check as fast as possible :>

File: openvdb/openvdb/points/AttributeArray.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmlait) [reply openvdb/openvdb/points/AttributeArray.cc:70]: My reading is it switches from a copy constructor to using an explicit .load()?

File: openvdb/openvdb/viewer/Viewer.cc (hunk 2 above)
Comment (jmlait) [reply openvdb/openvdb/viewer/Viewer.cc:333]: The main advantage of my proposal is it more closely matches the flow of the old code.  But you are the one submitting, so I'd prefer it be whatever you are more confident in :>

File: openvdb/openvdb/points/AttributeArray.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (BenFrantzDale) [reply openvdb/openvdb/points/AttributeArray.cc:70]: Yeah. The std atomics are less implicit and so don't have a copy c'tor, which is good, I think 'cause it makes stuff like this clearer.

File: openvdb/openvdb/viewer/Viewer.cc (hunk 2 above)
Comment (BenFrantzDale) [reply openvdb/openvdb/viewer/Viewer.cc:333]: I like your conciseness. I hadn't noticed what you noticed; I was just mechanically porting it.

I'll change it to 
```
bool expected = true;
if (mRedisplay.compare_exchange_strong(expected, false)) {
    // mRedisplay was true and was just changed to false
```
Or should I put this into the if? I forget what C++ version we target... That would look like
```
if (bool expected = true;
    mRedisplay.compare_exchange_strong(expected, false)) {
    // mRedisplay was true and was just changed to false
```

File: openvdb/openvdb/openvdb.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (BenFrantzDale) [reply openvdb/openvdb/openvdb.cc:48]: True. I guess I'm just thinking that with full-relaxed loading, it could have been `uninitialize`d a moment ago and we would incorrectly conclude that it's initialized. With acquire semantics it's still suspect, but at least a reasonable case should be correct. I'm not changing it to relaxed in my PR :-)

File: openvdb/openvdb/python/pyutil.h [hunk 6]
```diff
@@ -7,10 +7,10 @@
 #include "openvdb/openvdb.h"
 #include "openvdb/points/PointDataGrid.h"
 #include <boost/python.hpp>
-#include <tbb/mutex.h>
-#include <map> // for std::pair
-#include <string>
+#include <mutex>
 #include <sstream>
+#include <string>
+#include <utility> // for std::pair
```
Comment (BenFrantzDale) [inline openvdb/openvdb/python/pyutil.h:13]: This change from `<map>` to `<utility>` is a drive-by. I just noticed the `#include` didn't match the comment.

**Comment by danrbailey** [comment]
In case you were not aware - it looks like there are still a lot of failures in the CI checks at the moment - some tbb::mutex references not being replaced in the python module (openvdb/python/pyutil.h), this `if (bool expected = true;` initialization pattern is c++17 only and there are a bunch of new GCC10/Clang10 issues related to std::swap and using atomic copy constructor.

File: openvdb/openvdb/viewer/Viewer.cc (hunk 2 above)
Comment (jmlait) [reply openvdb/openvdb/viewer/Viewer.cc]: That would require C++17... I'm actually not sure what version we are at either, but C++17 is pretty new, so I'd avoid it unless we had a serious need (like constexpr if in templates!) rather than a slightly cleaner if :>

**Comment by BenFrantzDale** [comment]
> In case you were not aware - it looks like there are still a lot of failures in the CI checks at the moment - some tbb::mutex references not being replaced in the python module (openvdb/python/pyutil.h), this `if (bool expected = true;` initialization pattern is c++17 only and there are a bunch of new GCC10/Clang10 issues related to std::swap and using atomic copy constructor.

The init statement is easy. The swap is tricky... Let's see... it looks like tbb's atomic has
```
    operator value_type() const volatile {                // volatile qualifier here for backwards compatibility
        return  to_value<value_type>(
                __TBB_load_with_acquire( to_bits_ref(my_storage.my_value) )
        );
    }
```
and
```
atomic<T>& operator=( const atomic<T>& rhs ) {this->store_with_release(rhs); return *this;}
```
and
```
    value_type store_with_release( value_type rhs ) {
       //TODO: unify with store<release>
        __TBB_store_with_release( to_bits_ref(my_storage.my_value), to_bits(rhs) );
        return rhs;
    }
```
so in
```
template<typename T, Index Log2Dim>
inline void
LeafBuffer<T, Log2Dim>::swap(LeafBuffer& other)
{
    std::swap(mData, other.mData);
    std::swap(mOutOfCore, other.mOutOfCore); //< These are atomic<Index32>.
}
```
that swap line, `std::swap(mOutOfCore, other.mOutOfCore);', that it's complaining about must in tbb be doing
```
    atomic<Index32> tmp(std::move(x));
    x = std::move(y);
    y = std::move(tmp);
```
which I think means it's really essentially doing
```
    atomic<Index32> tmp(x.load(acquire));
    x.store(y.load(acquire), cst);
    y.store(tmp.load(acquire), cst);
```
so I think this is a good error to have. I can give a drop-in replacement although I do wonder what the correct thing to do is, and if the fact that swap can't be atomic is a latent bug.

**Comment by BenFrantzDale** [comment]
Also, it appears that the `tbb::atomic` is part of the ABI:
```
/// This helper class conditionally declares mOutOfCore as an atomic only if doing so
/// doesn't break ABI compatibility.
template<typename T>
struct LeafBufferFlags
{
#if OPENVDB_ABI_VERSION_NUMBER >= 5
    /// The type of LeafBuffer::mOutOfCore
    using type = std::atomic<Index32>;
    static constexpr bool IsAtomic = true;
#else // OPENVDB_ABI_VERSION_NUMBER < 5
    // These structs need to have the same data members as LeafBuffer.
    ...
#endif
```
so I think switching it out would require an ABI bump; is that correct?

**Comment by jmlait** [comment]
> Also, it appears that the `tbb::atomic` is part of the ABI:
> 
> so I think switching it out would require an ABI bump; is that correct?

Thank you for looking out for the ABI, it is very important we preserve this.  And we can't change it in the current 8 as it is locked down already, so would have to block in a 9 guard.

However, we discussed this actual issue and concluded that tbb::atomic and std::atomic are designed to go be bitwise compliant, so we believe we are ABI safe if we make this change.

(In some sense, we abandon all cross-compiler ABI at this point as std::atomic is not fully defined at the bit level, but that is something we are willing to live with as well)

**Comment by danrbailey** [comment]
Looks like this is making great progress! I see only a few outstanding CI checks failing due to use `Lock` and `Mutex` typedefs in the Houdini plugin which are going away:

https://github.com/AcademySoftwareFoundation/openvdb/pull/969/checks?check_run_id=1961374225#step:9:1024

Also - the DCO check is failing again, so we'll probably need to force push or rebuild this PR to fix that too...

**Comment by BenFrantzDale** [comment]
OK, I just pushed `openvdb_houdini` fixes.

**Comment by BenFrantzDale** [comment]
OK, I rebased so every commit should build and signed off. I left the comment about the ABI in as a single commit. Should I remove that comment?

This...
* Switches to using `std::atomic` directly.
* Switches to using `std::mutex` and `std::lock_guard<std::mutex>` directly, rather than using a `using` declaration (which was repeated in a few places).

**Review by danrbailey — APPROVED** [review]
This looks great, thanks for doing all this work!

File: openvdb/openvdb/unittest/TestTools.cc [hunk 7]
```diff
@@ -1600,21 +1600,22 @@ struct FloatToVec
...
-    FloatToVec() { numTiles = 0; }
+    FloatToVec() : numTiles{0} {}
+    FloatToVec(const FloatToVec& other) : numTiles{other.numTiles.load(std::memory_order_acquire)} {}
 
     void operator()(const InIterT& it, Accessor& acc)
     {
         if (it.isVoxelValue()) { // set a single voxel
             acc.setValue(it.getCoord(), toVec(*it));
         } else { // fill an entire tile
-            numTiles.fetch_and_increment();
```
Comment (jmlait) [inline openvdb/openvdb/unittest/TestTools.cc:1610]: You would think with ambiguous api calls like this it would be
fetch_then_increment
and
fetch_then_add
to make it clear that the result of the fetch is from before the operation in both cases.   The fetch_add official docs do not help with wonderful lines like

Return value
The value immediately preceding the effects of this function in the modification order of *this.

That was helpful!

**Review by jmlait — APPROVED** [review]

**Comment by Idclip** [comment]
Is this good to go? Are people only concerned about the ABI? Do we need to speak to the tbb devs? Otherwise lets get this in

**Comment by danrbailey** [comment]
> Is this good to go? Are people only concerned about the ABI? Do we need to speak to the tbb devs? Otherwise lets get this in

Yes, I believe so. I was hoping to find time to bring this in today, but didn't get to it. We need to merge this PR to a new branch, add release notes then create a new PR to merge into master. Think I saw a comment that needs deleting now we've answered the question about this being ABI-compatible too.

**Comment by JackBoosY** [comment]
Fine, let's waiting for the next release.
